diff --git a/dom/base/Element.cpp b/dom/base/Element.cpp
--- a/dom/base/Element.cpp
+++ b/dom/base/Element.cpp
@@ -3,6 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "mozilla/dom/Element.h"
+#include "SpeculumMutationObserver.h"
 
 #include "AnimationCommon.h"
 #include "CustomElementRegistry.h"
@@ -1562,6 +1563,8 @@ already_AddRefed<ShadowRoot> Element::AttachShadowWithoutNameChecks(
 
   // 13. Set element's shadow root to shadow.
   SetShadowRoot(shadowRoot);
+  SpeculumNotifyShadowAttached(this, shadowRoot);
 
   // Dispatch a "shadowrootattached" event for devtools if needed.
   if (MOZ_UNLIKELY(
